0.5.quality_control/0.jump_umap_analysis_with_cosmicqc.py

```diff
@@ -0,0 +1,698 @@
+                if col not in schema_names
+            ],
+        ]
+    ]
+
+    # Connect to DuckDB and read the parquet file directly
+    with duckdb.connect() as ddb:
+
+        # execute a query to extract sample of data from the file without reading it in full
+        df_features_with_cqc_outlier_data = ddb.execute(
+            f"""
+            WITH sampled_data AS (
+                SELECT *,
+                       ROW_NUMBER() OVER (PARTITION BY Metadata_Well ORDER BY RANDOM()) AS rn,
+                       COUNT(*) OVER (PARTITION BY Metadata_Well) AS group_count
+                FROM read_parquet('{merged_single_cells}')
+            )
+            SELECT * FROM sampled_data
+            INNER JOIN df_labeled_outliers_unique_cols AS joined_data ON
+                joined_data.Metadata_ImageNumber = sampled_data.Metadata_ImageNumber 
+                AND joined_data.Image_Metadata_Row = sampled_data.Image_Metadata_Row
+                AND joined_data.Image_Metadata_Site = sampled_data.Image_Metadata_Site
+                AND joined_data.Metadata_ObjectNumber = sampled_data.Metadata_ObjectNumber
+                AND joined_data.Metadata_Plate = sampled_data.Metadata_Plate
+                AND joined_data.Metadata_Well = sampled_data.Metadata_Well
+                AND joined_data.Image_TableNumber = sampled_data.Image_TableNumber
+            WHERE rn <= group_count * {sample_fraction}
+            """
+        ).df()
+
+    df_features_with_cqc_outlier_data.to_parquet(parquet_sampled_with_outliers)
+
+else:
+    df_features_with_cqc_outlier_data = pd.read_parquet(parquet_sampled_with_outliers)
+
+df_features_with_cqc_outlier_data
+# -
+
+# show our data value counts regarding outliers vs inliers
+df_features_with_cqc_outlier_data[
+    "analysis.included_at_least_one_outlier"
+].value_counts()
+
+# prepare data for normalization and feature selection
+# by removing cosmicqc and analaysis focused columns.
+df_for_normalize_and_feature_select = df_features_with_cqc_outlier_data[
+    # read feature names from cytotable output, which excludes
+    # cosmicqc-added columns.
+    parquet.read_schema(merged_single_cells).names
+]
+# show the modified column count
+len(df_for_normalize_and_feature_select.columns)
+
+# join JUMP metadata with platemap data to prepare for annotation
+df_platemap_and_metadata = pd.read_csv(
+    filepath_or_buffer=(
+        "s3://cellpainting-gallery/cpg0000-jump-pilot/source_4"
+        "/workspace/metadata/platemaps/2020_11_04_CPJUMP1/"
+        "platemap/JUMP-Target-1_compound_platemap.txt"
+    ),
+    sep="\t",
+).merge(
+    right=pd.read_csv(
+        filepath_or_buffer=(
+            "s3://cellpainting-gallery/cpg0000-jump-pilot/source_4"
+            "/workspace/metadata/external_metadata/"
+            "JUMP-Target-1_compound_metadata.tsv"
+        ),
+        sep="\t",
+    ),
+    left_on="broad_sample",
+    right_on="broad_sample",
+)
+df_platemap_and_metadata
```

Comment:
This is very interesting! I do believe the metadata has been saved to the repository but this is a new way of getting a file that I never thought of. Consider using the metadata that has been added to this repo if possible, I could be wrong.

Reply:
I think in the initial approach here I thought there might be data missing which we needed for Pycytominer work. I'll double check on this and follow up.

Reply:
I found that none of the files under the `reference_plate_data` dir include the `well_number` which is how we join the data to the profiles. I think specifically we need `well_position` and `control_type` in order to perform the Pycytominer annotation and normalization steps correctly. Going to leave things here as-is as a result.